### ISSUE TITLE
X11 dependency

### ISSUE THREAD

**Issue by woutervanh** [lucimber/meta-openjdk-temurin#22]
Sounded very promissing, but there are no headless runtimes foreseen in this repo. And I'm building for a non-X11 device

**Comment by sebveit** [comment]
Have you even tried to use this layer? Those binaries work just fine under desktop or server.
The dependencies are neccessary runtime dependencies or otherwise bitbake won't bake the recipe.
I'm using one of those prebuilt binaries on a single-board computer without a display and it works just fine.

Can you be more specific?

**Comment by woutervanh** [comment]
Yes, I tried it and was not working for me I thought. However, I just had a second look at the recipes and don't see what the issue could have been at first sight... Maybe I'm mistaken. Will try again later on

**Comment by sebveit** [comment]
Thanks for the reply. Your post got me thinking ... maybe it's possible to reduce the runtime dependencies even further. The Eclipse Temurin VMs for Alpine Linux look promising. Those VMs might ressemble of what you think of headless runtimes. I will have a further look into them but can't promise anything.

If you encounter any problems while using this layer feel free to post here. This might help others.

**Comment by woutervanh** [comment]
Trying it again, and this is my error:
ERROR: Nothing RPROVIDES 'libx11' (but /opt/yocto/sources/meta-openjdk-temurin/recipes-core/openjdk-17-jre/openjdk-17-jre_17.0.10+7.bb RDEPENDS on or otherwise requires it)
libx11 was skipped: missing required distro feature 'x11' (not in DISTRO_FEATURES)

**Comment by sebveit** [comment]
I don't know your specific setup, so I'm guessing. Somehow bitbake won't use the recipe for `libx11` in your build.

Maybe there's no recipe for it or something (a custom layer or local.conf) is prohibiting the use of it on your build.

At first, you should take a look at `bitbake-layers show-recipes | grep libx11:` if there is a recipe for libx11 in your setup. It will print `libx11:` if there is a recipe for it. Take a look at https://docs.yoctoproject.org/dev/dev-manual/layers.html#managing-layers for more information on that topic.

If the above command doesn't work, you first have to source into your build directory via `source oe-init-build-env <path-to-your-build-dir>` from the root of your yocto or openembedded setup and then try again. 

To find out, if the recipe can be used, you need to take a look at your `DISTRO_FEATURES` by issuing `bitbake -e openjdk-17-jre | grep ^DISTRO_FEATURES`. This command will only print your effective settings used in your build. No building is going to happen. The first output is of interest and should contain the string `x11`. Visit https://docs.yoctoproject.org/ref-manual/features.html#ref-features-distro for an explanation of those features.

Maybe it might help, if you try a vanilla setup and adapt your specific setup accordingly. You can take a look at the files of this project:
* Use https://github.com/lucimber/meta-openjdk-temurin/blob/main/.github/workflows/bitbake-main-local.conf as your local.conf
* and follow the steps from https://github.com/lucimber/meta-openjdk-temurin/blob/main/.github/workflows/bitbake-main.yml to setup your vanilla build

**Comment by woutervanh** [comment]
Well, the issue is not that I cannot foresee X11, the issue is that I don't understand why this is a dependency. I'm not planning to do any X11 stuff with it, and my embedded system does not have X11 at all (and don't need it). So just to run a basic java application, I need a lot of unnecessary (as in I don't need it) stuff on an already limited system.  But I'm building it now with x11 enabled, just to try it out

**Comment by sebveit** [comment]
TLDR: You have to supply runtime dependencies, that get baked into your image, wether you like it or not. Sorry!

Those prebuilt JVMs offered by Adoptium and used by this layer are built for general purpose and therefore also contain code that is used when doing a GUI (java.awt.* packages). Bitbake analysis those binaries and complains and even stops the process of building your embedded image when the recongnized runtime dependencies of those binaries aren't satisfied by your build setup (e.g. via RDEPENDS in the recipe of openjdk-xx-jre).

That's why you have to supply those runtime dependencies even when you're not going to use them. It won't built the whole X11 server, just the library on which the OpenJDK VM depends on. Yes, this isn't optimal but the additional disk usage shouldn't bother because it isn't that much.

It simply boils down to a tradeoff between having the newest and tested JVM vs. reduced space consumption and less libraries (runtime dependencies) on your device. If you want less runtime dependencies you must use the VM (OpenJDK 8) from the official layer (meta-java), but then you're stuck with a rather untested outdated Java 8 VM.

I'll even bet, that there are recipes with unecessary (read: features you don't need by the resulting binary) runtime dependencies, but get baked into your image anyway. Of course the people from Yocto/OE try their best to minimize the resulting dependency hell but that isn't always going to work for each recipe/software. It's up to the developer of the specific software (e.g. library, tool or VM) to reduce the runtime dependencies of the resulting binary. Either by baking the dependency into the binary, which makes it larger, or make the dependcy optional so it can be left out.

**Comment by woutervanh** [comment]
Agree, I just had a quick smoke test installing the x11-libs along, and now it's working fine. Thanks!

I was under the impression you needed to run an X11 system, but this is not the case. I just need the libraries available. I still have to figure out the size of the additional libs, but compared to the complete openjdk blob, that will be quiet limited indeed.
Will close the issue as won't fix.

**Comment by Daftus** [comment]
Hello,

I'm trying to build an imx-image-multimedia image and I'm encountering the same issue with the x11 dependency. The correct package libX11-1.7.3.1 is available in the downloads folder but I keep getting the same error:

`ERROR: Nothing RPROVIDES 'libx11' (but /opt/yocto-build/imx-yocto-bsp/sources/meta-openjdk-temurin/recipes-core/openjdk-17-jre/openjdk-17-jre_17.0.10+7.bb RDEPENDS on or otherwise requires it) 
libx11 was skipped: missing required distro feature 'x11' (not in DISTRO_FEATURES)`

I even added the `DISTRO_FEATURES:append = " x11"` in my local.conf

I'm out of solution, do you have an idea ?
Thank you

**Comment by sebveit** [comment]
Hello Daftus,

I'm not familiar with imx-images, but you should look at the readme at https://github.com/nxp-imx/meta-imx. It points out the following under section **Build images**:
> The X11 and Framebuffer distros are only supported for i.MX 6 and i.MX 7.  i.MX 8 should use xwayland only.

Have you tried the command `bitbake -e openjdk-17-jre | grep ^DISTRO_FEATURES` to find out if it's the case for you?

**Comment by Daftus** [comment]
Hello Sebastian,

First of all thank your for the quick answer ! Secondly, I was actually able to generate an image by switching my DISTRO from `fsl-imx-wayland` to `fsl-imx-x11` (makes sense). The OpenJDK was indeed present in the image after all of this.

I don't know if it should work for a wayland distro but I can say it is not !
On another note, I recommend adding the part of your answer from this [OpenJDK17 Yocto](https://stackoverflow.com/questions/75823654/run-java-17-on-yocto-os) to your READ.ME, especially the part where you indicate to add `openjdk-17-jre`.

Thank you !

**Comment by woutervanh** [comment]
I was able to build it using:
DISTRO_FEATURES:remove = "directfb "
DISTRO_FEATURES:append = " wayland x11"

**Comment by Daftus** [comment]
Good to know thank you for the feedback !

**Comment by stobias123** [comment]
Hmm looking at this same issue, trying to find the headless version (doesn't require X11). Not because we _cant_ find x11 libs, but we dont want to include them. Does adoptium publish headless binaries? I'm having trouble finding them.

**Comment by jdelbefoss** [comment]
Hello, and thank you already for this layer!

Would it be possible to reduce the scope of these libx* RDEPENDS to only the components that really need them?
For instance by adding more sub-packages, maybe together with #57 ?

**Comment by sebveit** [comment]
I share the wish for a headless version!
As of now, I don't know of any headless binaries published by adoptium or a similar project.
The take with the reduced scope is nice, but I don't know if that's possible.

Since I'm the maintainer of this project and share the same thoughts on this,
I'll look regularly into this problem hoping to find a useful solution.

If someone finds a way to takle this, please reach out! thanks

**Comment by jdelbefoss** [comment]
So I have tried the following with the updated scarthgap branch, in openjdk-21-jdk-native_21.0.4+7.bb:

```
JVM_RDEPENDS:remove:x86-64 = "\
	alsa-lib \
	libx11 \
	libxext \
	libxi \
	libxrender \
	libxtst \
"
```

In another native recipe that DEPENDS on it, if I give the correct JAVA_HOME (${STAGING_LIBDIR_NATIVE}/jvm/openjdk-21-jdk), it seems that I'm able to use (some of) the java utils in a headless build.
I have not tested much, but this is already enough for my use case!
